awx/main/scheduler/task_manager.py

```diff
@@ -434,6 +428,7 @@ def generate_dependencies(self, task):
                 latest_project_update = self.get_latest_project_update(task)
                 if self.should_update_related_project(task, latest_project_update):
                     project_task = self.create_project_update(task)
+                    self.process_pending_tasks([project_task])
```

Comment:
The only thing I don't like about this is the potential recursion. I get that you're changing the call argument, but still.. we're here in `generate_dependencies`, which is called by `process_pending_tasks`. Then it calls `process_pending_tasks`.

It would be ideal to restructure code in a way that it obviously does not recurse. But for starters, I want a justification why this is believed to be a safe call to a method further up the call stack.

Reply:
(answering my own question)

because this `project_task` is newly created, it shouldn't be in the `pending_tasks` lists that the original `process_pending_tasks` is looping over. When you create a new dependent task, you are making an effort here to have it jump the line, get processed before the task that depends on it.

Reply:
good points.

yeah we need to skip ahead mostly so that add_job() is called on the project update and added to dependency graph. That way when we get back to the main loop of pending jobs, is_job_blocked() will return True (because now it's waiting on the newly generated project update) and delay the start.

the recursion shouldn't be an issue :crossed_fingers:  because generate_dependencies is only for type Job and will return instantly for ProjectUpdate and InventoryUpdate types.

Reply:
and in the case where add_job() is not called for the project update (i.e. can't find an instance to run on), the job still won't run because the last portion of is_job_blocked() will do a task.dependent_jobs_finished() check and return True if it's project update hasn't finished.